lib/segment/src/index/field_index/geo_index.rs

```diff
@@ -1 +1,531 @@
+use crate::types::{GeoBoundingBox, GeoPoint, GeoRadius};
+use geo::Coordinate;
+use geohash::{decode, encode, neighbor, Direction};
+use std::ops::Range;
 
+type GeoHash = String;
+
+/// Max size of geo-hash used for indexing. size=12 is about 6cm2
+const GEOHASH_MAX_LENGTH: usize = 12;
+
+/// Size of geo-hash grid depending of the length of hash.
+/// Format:
+///   For hash length: N = 4
+///   let (lat_length_in_meters, long_length_in_meters) = HASH_GRID_SIZE[N - 1];
+/// Source: <https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations-bucket-geohashgrid-aggregation.html#_cell_dimensions_at_the_equator>
+const HASH_GRID_SIZE: &[(f64, f64)] = &[
+    (5009.4 * 1000.0, 4992.6 * 1000.0),
+    (1252.3 * 1000.0, 624.1 * 1000.0),
+    (156.5 * 1000.0, 156.0 * 1000.0),
+    (39.1 * 1000.0, 19.5 * 1000.0),
+    (4.9 * 1000.0, 4.9 * 1000.0),
+    (1.2 * 1000.0, 609.4),
+    (152.9, 152.4),
+    (38.2, 19.),
+    (4.8, 4.8),
+    (1.2, 0.595),
+    (0.149, 0.149),
+    (0.037, 0.019),
+];
+
+impl From<GeoPoint> for Coordinate<f64> {
+    fn from(point: GeoPoint) -> Self {
+        Self {
+            x: point.lat,
+            y: point.lon,
+        }
+    }
+}
+
+#[derive(Debug)]
+struct FullGeoBoundingBox {
+    north_west: GeoPoint,
+    south_west: GeoPoint,
+    south_east: GeoPoint,
+    north_east: GeoPoint,
+}
+
+impl FullGeoBoundingBox {
+    fn from_geo_bounding_box(bounding_box: &GeoBoundingBox) -> Self {
+        let GeoPoint {
+            lat: max_lat,
+            lon: min_lon,
+        } = bounding_box.top_left;
+        let GeoPoint {
+            lat: min_lat,
+            lon: max_lon,
+        } = bounding_box.bottom_right;
+
+        let north_west = GeoPoint {
+            lon: min_lon,
+            lat: max_lat,
+        };
+        let south_west = GeoPoint {
+            lon: min_lon,
+            lat: min_lat,
+        };
+        let south_east = GeoPoint {
+            lon: max_lon,
+            lat: min_lat,
+        };
+        let north_east = GeoPoint {
+            lon: max_lon,
+            lat: max_lat,
+        };
+
+        Self {
+            north_west,
+            south_west,
+            south_east,
+            north_east,
+        }
+    }
+
+    fn shortest_side_length_in_km(&self) -> f64 {
```

Comment:
Would it make sense to use https://crates.io/crates/uom ? Or do you think it is an overkill here?